File: src/main/java/com/hazza/LeetCodeByJava/No3LongestSubstringWithoutRepeatingCharacters.java
package com.hazza.LeetCodeByJava;

/**
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/#/description
 */
public class No3LongestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {
        boolean[] flag = new boolean[128];
        int maxLen = 0;
        int len = 0;
        for (int i = 0;i < s.length();++i) {
            char c = s.charAt(i);
            if (flag[c] == true) {
                flag = new boolean[128]; //建立一个ASCII码值表
                maxLen = (maxLen > len) ? maxLen : len;
                i -= len; //若出现重复字符，恢复到重复字符第一次出现的地方，并在下一次循环加1，恢复到重复字符第一次出现的下一个字母
                len = 0;
            } else {
                ++len;
                flag[c] = true;
            }
        }
        maxLen = (maxLen > len) ? maxLen : len; //防止出现没有重复字符的字符串

        return maxLen;
    }
}

File: src/main/python/com.hazza.LeetCodeByPython/SegmentOne/No3_Longest_Substring_Without_Repeating_Characters.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/3 19:31
# https://leetcode.com/problems/longest-substring-without-repeating-characters/


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxLen = 0
        start = i = 0
        used = {}
        for i in range(len(s)):
            if s[i] in used and start <= used[s[i]]:
                start = used[s[i]] + 1
            else:
                maxLen = max(maxLen, i - start + 1)
            used[s[i]] = i

        return maxLen


if __name__ == '__main__':
    solution = Solution()
    assert solution.lengthOfLongestSubstring("abcabcbb") == 3
    assert solution.lengthOfLongestSubstring("bbbbb") == 1
    assert solution.lengthOfLongestSubstring("pwwkew") == 3